Modify content security policy

Henni opened the pull request:
fixes #47 

~~probably breaks compatibility for owncloud < 8.1~~

DJaeger commented:
Because of the compatibility break I have asked before I have created a PR.
If a legacy version should be preserved, another branch should be created before.

DJaeger commented:
For fixing
```
Content Security Policy: Die Einstellungen der Seite haben das Laden einer Ressource auf self blockiert ("script-src https://my.url 'unsafe-eval'").
onsubmit attribute on DIV element
```
I think you have to also add
```
		$csp->addAllowedScriptDomain('\'self\'');
```

brantje commented:
Nice, thanks for your contribution!

Henni commented:
I added `$csp->addAllowedScriptDomain('\'self\'');` and made the code backwards compatible like @LukasReschke described at https://github.com/owncloud/maps/issues/47#issuecomment-129461851.

v1r0x commented:
For me this only works, if I add

```
$csp->addAllowedImageDomain('*');
```

Henni commented:
@v1r0x What errors are shown with the current implementation?

v1r0x commented:
```
Content Security Policy: Die Einstellungen der Seite haben das Laden einer Ressource auf self blockiert ("script-src https://owncloud.my.url https://owncloud.my.url 'unsafe-eval'").
onsubmit attribute on DIV element

Content Security Policy: Die Einstellungen der Seite haben das Laden einer Ressource auf http://otile3.mqcdn.com/tiles/1.0.0/osm/15/17210/11312.png blockiert ("img-src https://owncloud.my.url https://otile1.mqcdn.com https://otile2.mqcdn.com https://otile3.mqcdn.com https://otile4.mqcdn.com").
Content Security Policy: Die Einstellungen der Seite haben das Laden einer Ressource auf http://otile2.mqcdn.com/tiles/1.0.0/osm/15/17210/11311.png blockiert ("img-src https://owncloud.my.url https://otile1.mqcdn.com https://otile2.mqcdn.com https://otile3.mqcdn.com https://otile4.mqcdn.com").
Content Security Policy: Die Einstellungen der Seite haben das Laden einer Ressource auf http://otile2.mqcdn.com/tiles/1.0.0/osm/15/17209/11312.png blockiert ("img-src https://owncloud.my.url https://otile1.mqcdn.com https://otile2.mqcdn.com https://otile3.mqcdn.com https://otile4.mqcdn.com").
```

Henni commented:
@v1r0x did you use the master branch or did you checkout this pull-request when these errors occurred?

v1r0x commented:
I checked out the pull request

Henni commented:
@v1r0x which browser did you use?

brantje commented:
```
$csp->addAllowedImageDomain('www.openstreetmap.org');
```

DJaeger commented:
@brantje: The blocked images doesn't came from osm but from mq, so this doesn't work.

The issue seems to be, that the images to load use http and allowed is only https.
The csp class seems to prepend the currently used protokoll and @v1r0x seems to use https to access his owncloud.
But we have to force http as mq only delivers tiles over http.

DJaeger commented:
As of 
https://developer.mozilla.org/en-US/docs/Web/Security/CSP/CSP_policy_directives
we can also allow
http://*.mqcdn.com
or
http://otile1.mqcdn.com
http://otile2.mqcdn.com
http://otile3.mqcdn.com
http://otile4.mqcdn.com

DJaeger commented:
As the csp class simply pass through the values this should work.

Henni commented:
@DJaeger I applied your fix `http://*.mqcdn.com`

@v1r0x could you retest please.

v1r0x commented:
works fine now! :)
But I had to add
```php
$csp->addAllowedImageDomain('https://api.tiles.mapbox.com');
```
as well for the markers

DJaeger commented:
@v1r0x: You are right, they are also requested from another origin.
I think they could be added to this repo some time later, as they can be found here:
https://github.com/mapbox/maki
This would minimize dependencies and requests to other origins and can also be handed by the caching preferences of owncloud.

Henni commented:
For now I'll add `https://api.tiles.mapbox.com` to the Content-Security-Policies.

Henni commented:
If you're happy with these changes I'll squash the commits.

Henni commented:
I created an issue to integrate the icons: #52

DJaeger commented:
Forcing https for the API is great, as it is hardcoded to https.
Adding the specific subdomains would be more secure, but I think we can trust a subdomain wildcard for this domain enough to let it as it is.
The third code comment is wrong, as it allows scripts to be loaded and not images.

Henni commented:
I didn't completely understand the last rule. As far as I understand it was added because some image which is included in the script couldn't load. What would be a better comment in your opinion @DJaeger?

brantje commented:
Looks good. :+1:

jancborchardt commented:
Nice, looks great! :) @Henni cool to see you in here as well!

@LukasReschke @DJaeger can you give this a review too please? So with this Maps will only be compatible with 8.1+? (I’m fine with that, just confirming.)

Henni commented:
@jancborchardt it should still be compatible for < 8.1
maybe @LukasReschke can confirm this.

DJaeger commented:
An image can only be in a script as "data:" url, where it would require 
```
$csp->addAllowedImageDomain('data:');
```
And as I have now seen, we also require this for contact images.

```
$csp->addAllowedScriptDomain('\'self\'');
```
is required to load scripts by scripts from the server.
I don't know where we do it, but @v1r0x had such a request, that was blocked.
I'm not sure, but it could be coursed by wrongly requesting json.

DJaeger commented:
@jancborchardt: If I'm right it should be compatible with each oc8 version

Henni commented:
@DJaeger what about oC7?

DJaeger commented:
I'm sorry, I was wrong.
We have not changed any compatibility with this PR.

jancborchardt commented:
So we merge? :+1: from me :)

DJaeger commented:
No, please add this line before:
```
$csp->addAllowedImageDomain('data:');
````

DJaeger commented:
And change 
```
// images inside scripts
```
to
```
// scripts
```

jancborchardt commented:
@Henni and maybe as a last thing squash the commits into one to keep history clean. :)

DJaeger commented:
If these changes are done also :+1: from me

LukasReschke commented on controller/pagecontroller.php:
```diff
@@ -40,7 +40,18 @@ public function __construct($appName, IRequest $request, $userId, $cacheManager,
+			$csp->addAllowedImageDomain('http://*.mqcdn.com');
+			// marker icons
+			$csp->addAllowedImageDomain('https://api.tiles.mapbox.com');
+			// images inside scripts
```
Needs to get rephrased. You do allow any inline JS like `<script>alert(1)</script>`, which btw. makes CSP useless at all. But change the comment ;)

DJaeger replied:
We should better see, what exactly makes inline script execution required.

LukasReschke replied:
That would be a better solution. Yes.

Henni replied:
I don't completely understand this yet...
@LukasReschke Is `$csp->addAllowedScriptDomain('\'self\'');` the reason `<script>alert(1)</script>` is possible?
Because I believe this is enabled by default either way (see [contentsecuritypolicy.php](https://github.com/owncloud/core/blob/a9d1a0144018e60ba2728708bf965b4d9855920b/lib/public/appframework/http/contentsecuritypolicy.php#L36))

LukasReschke replied:
Ah. Sorry. I misread it. I thought this was `allowInlineScript`… :speak_no_evil: 

This is not needed at all then. What you likely need is that as per https://github.com/owncloud/maps/pull/49#issuecomment-129423324. `onsubmit` is inline JS.

Henni replied:
@LukasReschke ok thanks. I just removed this line but I won't add `allowInlineScript`.

LukasReschke commented on controller/pagecontroller.php:
```diff
@@ -40,7 +40,18 @@ public function __construct($appName, IRequest $request, $userId, $cacheManager,
+			// map tiles
+			$csp->addAllowedImageDomain('http://*.mqcdn.com');
+			// marker icons
+			$csp->addAllowedImageDomain('https://api.tiles.mapbox.com');
```
better go with `*` to all image domains as CSP is anyways disabled below :-)

LukasReschke replied:
disabled === provides no protection

DJaeger replied:
Why allow all image domains?

LukasReschke replied:
If you are sure that it are only these: You can keep them as-is. However, a lot of the protection offered by CSP is anyways gone due through https://github.com/owncloud/maps/pull/49#discussion_r36758549. – Shouldn't make things weaker.

DJaeger replied:
It are only these I have posted.
If there are some missing, we can add them later.
We should not make this unsafe just becourse the line below makes it unsafe.
We have anyway some rework to do. ;-)

LukasReschke commented:
Valid approach. I'd make it easier by just whitelisting `*` and also change the comment. Besides this: Good work :+1:

Henni commented:
I removed `$csp->addAllowedScriptDomain('\'self\'');` as this is set by default.
I added `$csp->addAllowedImageDomain('data:');` as it is needed for inline images.

To sum it up, we now allow:
* Map tiles (on four different subdomains) - `addAllowedImageDomain('http://*.mqcdn.com');`
* Map icons (which will change with #52) - `addAllowedImageDomain('https://api.tiles.mapbox.com');`
* Inline images (for loading tile and contact thumbnails when fixes) - `addAllowedScriptDomain('data:');`

brantje commented:
:+1: code looks good

DJaeger commented:
:+1: from me.

Henni commented:
@brantje @DJaeger great thanks!
